### PR TITLE
feat(react): add `BaseChat` and `ThreadChat` components

### LINKED ISSUE #3897: Chat React Component

### PR DESCRIPTION
**Pull request by ThatOneBro**
Closes #3897

### CODE REVIEW

**Review by codyebberson — COMMENTED** [review]
Overall components look good

We should have a brainstorming session on component naming.  `Chat` and `ChatBox` feels at risk of confusion. (I personally wouldn't be able to infer the difference without looking at the code)

`Chat` vs `ThreadChat` ?

**Comment by ThatOneBro** [comment]
> `Chat` vs `ThreadChat` ?

What do you think about `BaseChat` vs `ThreadChat`? I think `Chat` might drive a lot of people to attempt to use `Chat` without realizing it probably is not the choice for net new stuff

**Comment by rahul1** [comment]
`BaseChat` vs `ThreadChat` sounds reasonable to me. Indicates that `BaseChat` is a lower-level primitive

**Comment by rahul1** [comment]
Request for storybook stories as well 🙏

File: examples/medplum-live-chat-demo/src/pages/HomePage.tsx [hunk 1]
```diff
@@ -22,33 +23,112 @@ export function HomePage(): JSX.Element {
   // https://www.medplum.com/docs/tutorials/register
   const profile = useMedplumProfile() as Practitioner;
   const medplum = useMedplum();
+  const [thread, setThread] = useState<Communication>();
+
+  const profileRef = useMemo(() => createReference(profile as ProfileResource), [profile]);
```
Comment (codyebberson) [inline examples/medplum-live-chat-demo/src/pages/HomePage.tsx:28]: Super minor, but these `createReference` and `getReferenceString` calls probably do not need `useMemo` - they're all pretty cheap.

File: packages/react/src/BaseChat/BaseChat.test.tsx [hunk 2]
```diff
@@ -0,0 +1,338 @@
+import { ProfileResource, createReference, getReferenceString } from '@medplum/core';
+import { Bundle, Communication } from '@medplum/fhirtypes';
+import { BartSimpson, DrAliceSmith, HomerSimpson, MockClient } from '@medplum/mock';
+// @ts-expect-error _subscriptionController is not exported from module normally
```
Comment (codyebberson) [inline packages/react/src/BaseChat/BaseChat.test.tsx:4]: Hm, any way we could avoid this?

File: packages/react/src/BaseChat/BaseChat.test.tsx [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+
+const homerReference = createReference(HomerSimpson);
+const homerReferenceStr = getReferenceString(homerReference);
+const drAliceReference = createReference(DrAliceSmith);
+const drAliceReferenceStr = getReferenceString(drAliceReference);
+const HOMER_DR_ALICE_CHAT_QUERY = `sender=${homerReferenceStr},${drAliceReferenceStr}&recipient=${homerReferenceStr},${drAliceReferenceStr}`;
+
+jest.mock('@medplum/react-hooks', () => {
+  // eslint-disable-next-line @typescript-eslint/no-var-requires
+  const EventEmitter = require('node:events');
```
Comment (codyebberson) [inline packages/react/src/BaseChat/BaseChat.test.tsx:23]: Does this have to be `EventEmitter` from `node:events`?

Could we use the `EventTarget` from `@medplum/core`?

File: packages/react/src/ResourceBlame/ResourceBlame.module.css [hunk 4]
```diff
@@ -13,7 +13,7 @@
   & td {
     padding: 2px 4px 0 4px;
     vertical-align: top;
-    white-space: nowra;
+    white-space: nowrap;
```
Comment (codyebberson) [inline packages/react/src/ResourceBlame/ResourceBlame.module.css:16]: 😬  nice catch

**Review by codyebberson — APPROVED** [review]
Nice.

Minor non-blocking comments.

2 of the sonar issues could be fixed by adding a quick comment to the empty body

Kudos for jumping through all of the hoops to make testing work.

**Pull request merged**